### DISCUSSION TITLE
Question to understand data flow in TodosViewModel

### DISCUSSION THREAD

**Discussion by magicwave** [pointfreeco/swift-identified-collections#76] [Q&A]
I've just read [Introducing: identified collections](https://github.com/pointfreeco/swift-identified-collections?tab=readme-ov-file#introducing-identified-collections), where example code is given of a function, `func todoCheckboxToggled(at id: Todo.ID) async` that mutates the state of an element in the published `todos` `IdentifiedArray.` 

The first line in the function, `self.todos[id: id]?.isComplete.toggle()`, appears to directly mutate the todo element state, which I assume will cause the SwiftUI view driven by the viewModel to update. I understand this as mutating local viewModel data for the purpose of immediately updating the todo list view?

Next is a `do/try/catch` block where presumably persistent `todo` data is mutated asynchronously via an `apiClient` instance. 

Further, I assume that the viewModel implements an observer for changes to the persistent data, something along the lines of:
```
init() {
    self.apiClient.$publishedToDos
        .assign(to: &$todos)
}
```

With these assumptions, I wonder why we toggle local todo state in addition to updating the (again presuming) more canonical data source via `apiClient`? In that scenario it seems we have two sources of truth: local and persistent. I might expect that we would not change the local data source in the `ToDoViewModel` toggle function, and simply update, and let the list view react to, the change in the persistent data source via the publisher.

**Comment by kamcma** [comment]
> Further, I assume that the viewModel implements an observer for changes to the persistent data, something along the lines of:
> 
> ```
> init() {
>     self.apiClient.$publishedToDos
>         .assign(to: &$todos)
> }
> ```
[This is a big assumption](https://en.wikipedia.org/wiki/Assume_a_can_opener)! For the purposes of the example, I think it's far more reasonable to assume `apiClient` is a simple CRUD HTTP client with no observability mechanism.

The problem of keeping local memory state and remote server state in sync is simply a higher-level problem than what this library exists to solve, and I don't think the example code was attempting to comment on it.

Edit: your question seems to assume `apiClient` is holding onto todos internally. That would actually be _three_ sources of the truth: the remote API, whatever `apiClient` is holding onto internally, and the `TodosViewModel`. The example really only assumes the remote API and `TodosViewModel`, and `apiClient` is more stateless. And again: keeping those in sync is outside the scope of the library.
